lsp/server/src/diagnostic/gql/over-sized-record.ts

```diff
@@ -0,0 +1,167 @@
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { DiagnosticProducer } from '../DiagnosticProducer';
+import { Diagnostic, DiagnosticSeverity } from 'vscode-languageserver/node';
+import type { ASTNode, FieldNode } from 'graphql';
+
+import {
+    generateEntityTree,
+    resolveEntityNameFromMetadata,
+    getFieldSize
+} from '../../utils/gqlUtils';
+import { OrgUtils } from '../../utils/orgUtils';
+import type { RootNode, EntityNode } from '../../utils/gqlUtils';
+
+const MAX_ALLOWED_SIZE = 32768;
+
+export const OVER_SIZED_FIELD_MESSAGE =
+    'This field’s value could exceed 32 KB. Large data sizes can negatively affect mobile app performance, and potentially result in fewer returned records than expected.';
+export const OVER_SIZED_RECORD_MESSAGE =
+    'The total data size of all record fields could exceed 32 KB. Large data sizes can negatively affect mobile app performance, and potentially result in fewer returned records than expected.';
+
+const SEVERITY = DiagnosticSeverity.Information;
+
+export const RULE_ID = 'over-sized-record';
+export class OversizedRecord implements DiagnosticProducer<ASTNode> {
+    async validateDocument(
+        textDocument: TextDocument,
+        rootNode: ASTNode
+    ): Promise<Diagnostic[]> {
+        const rootDiagnosticNode = generateEntityTree(rootNode);
+        const rawDiagNodes = await createDiagnostics(rootDiagnosticNode);
+
+        const { overSizedEntities, overSizedFields } = rawDiagNodes;
+
+        return [
+            ...overSizedFields.map((rawNode) => {
+                return createLspDiagnostic(
+                    textDocument,
+                    rawNode,
+                    OVER_SIZED_FIELD_MESSAGE
+                );
+            }),
+            ...overSizedEntities.map((rawNode) => {
+                return createLspDiagnostic(
+                    textDocument,
+                    rawNode,
+                    OVER_SIZED_RECORD_MESSAGE
+                );
+            })
+        ];
+    }
+
+    getId(): string {
+        return RULE_ID;
+    }
+}
+
+export interface OverSizedDiagnostics {
+    overSizedFields: Array<FieldNode>;
+    overSizedEntities: Array<FieldNode>;
+}
+export async function createDiagnostics(
+    rootNode: RootNode
+): Promise<OverSizedDiagnostics> {
+    const results: OverSizedDiagnostics = {
+        overSizedFields: [],
+        overSizedEntities: []
+    };
+
+    for (const operationNode of rootNode.operations) {
+        for (const entityNode of operationNode.entities) {
+            try {
+                await generateDiagnostic(entityNode, results);
+            } catch (e) {
+                console.log(
+                    `Cannot conduct over-sized record diagnostic for ${entityNode.name} `
+                );
+            }
+        }
```

Comment:
Suppose `generateDiagnostic` throws where `results` from previous iteration has already accumulated elements. Then in this case is it an expected behavior that `results` may contain diagnostics? Or should this function be returning empty array?

Reply:
Good question. Each EntityNode is an independent unit from its peers. Error in one Entity does not interfere with its peer. So iteration should continue if one entityNode processing produces error.

Reply:
Can we also log `e.message` please.

Reply:
added